killswitch.py

```diff
@@ -1,54 +1,231 @@
+
+        else:
+            raise ValueError(f'Dont know how to index a {type(current)} ({current!r})')
+
+    _apply(current, path, to_set, delete)
+
+
+def _get_int(s: str) -> Optional[int]:
+    try:
+        return int(s)
+    except ValueError:
+        return None
+
+
+class KillSwitches(NamedTuple):
     """One version's set of kill switches."""
 
     version: semantic_version.SimpleSpec
-    kills: Dict[str, str]
+    kills: Dict[str, SingleKill]
+
+    @staticmethod
+    def from_dict(data: KillSwitchSetJSON) -> KillSwitches:
+        """Create a KillSwitches instance from a dictionary."""
+        ks = {}
+
+        for match, ks_data in data['kills'].items():
+            ks[match] = SingleKill(
+                match=match,
+                reason=ks_data['reason'],
+                redact_fields=ks_data.get('redact_fields'),
+                set_fields=ks_data.get('set_fields'),
+                delete_fields=ks_data.get('delete_fields')
+            )
+
+        return KillSwitches(version=semantic_version.SimpleSpec(data['version']), kills=ks)
 
 
 class DisabledResult(NamedTuple):
     """DisabledResult is the result returned from various is_disabled calls."""
 
     disabled: bool
-    reason: str
+    kill: Optional[SingleKill]
+
+    @property
+    def reason(self) -> str:
+        """Reason provided for why this killswitch exists."""
+        return self.kill.reason if self.kill is not None else ""
+
+    def has_kill(self) -> bool:
+        """Return whether or not this DisabledResult has a Kill associated with it."""
+        return self.kill is not None
+
+    def has_rules(self) -> bool:
+        """Return whether or not the kill on this Result contains rules."""
+        # HACK: 2021-07-09 # Python/mypy/pyright does not support type guards like this yet. self.kill will always
+        # be non-None at the point it is evaluated
+        return self.has_kill() and self.kill.has_rules  # type: ignore
 
 
 class KillSwitchSet:
     """Queryable set of kill switches."""
 
-    def __init__(self, kill_switches: List[KillSwitch]) -> None:
+    def __init__(self, kill_switches: List[KillSwitches]) -> None:
         self.kill_switches = kill_switches
 
-    def get_disabled(self, id: str, *, version=_current_version) -> DisabledResult:
+    def get_disabled(self, id: str, *, version: Union[Version, str] = _current_version) -> DisabledResult:
         """
         Return whether or not the given feature ID is disabled by a killswitch for the given version.
 
         :param id: The feature ID to check
         :param version: The version to check killswitches for, defaults to the current EDMC version
         :return: a namedtuple indicating status and reason, if any
         """
+
+        if isinstance(version, str):
+            version = semantic_version.Version.coerce(version)
+
         for ks in self.kill_switches:
             if version not in ks.version:
```

Comment:
I had to look up the semantic_version docs *and* run a test to be sure, so add a comment here that, yes, this 'in' syntax does support things like '<1.2.3' for range comparisons.

Reply:
NB: This could be in Killswitches.md#versions rather than in the code.  All it does is say it's a SimpleSpec, and examples are simple specific versions, not comparison versions.  We need to document that range comparison versions are supported.

Reply:
Clarified in docs